mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClustering.scala

```diff
@@ -0,0 +1,627 @@
+      centers.toBreeze.map(elm => elm + random.nextDouble() * elm * this.randomRange)
+    )
+  }
+
+  /**
+   * Splits the given cluster (tree) with bi-sect k-means
+   *
+   * @param clusterTree the splitted cluster
+   * @return an array of ClusterTree. its size is generally 2, but its size can be 1
+   */
+  private def split(clusterTree: ClusterTree): Array[ClusterTree] = {
+    val startTime = System.currentTimeMillis()
+    val data = clusterTree.data
+    val sc = data.sparkContext
+    var centers = takeInitCenters(clusterTree.center)
+
+    // TODO Supports distance metrics other Euclidean distance metric
+    val metric = (bv1: BV[Double], bv2: BV[Double]) => breezeNorm(bv1 - bv2, 2.0)
+    sc.broadcast(metric)
+
+    // If the following conditions are satisfied, the iteration is stopped
+    //   1. the relative error is less than that of configuration
+    //   2. the number of executed iteration is greater than that of configuration
+    //   3. the number of centers is equal to one. if one means that the cluster is not splittable
+    var numIter = 0
+    var error = Double.MaxValue
+    while (error > this.epsilon
+        && numIter < this.subIterations
+        && centers.size > 1) {
+      val startTimeOfIter = System.currentTimeMillis()
+
+      sc.broadcast(centers)
+      val newCenters = data.mapPartitions { iter =>
+        // calculate the accumulation of the all point in a partition and count the rows
+        val map = scala.collection.mutable.Map.empty[Int, (BV[Double], Int)]
+        iter.foreach { point =>
+          val idx = ClusterTree.findClosestCenter(metric)(centers)(point)
+          val (sumBV, n) = map.get(idx)
+              .getOrElse((new BSV[Double](Array(), Array(), point.size), 0))
+          map(idx) = (sumBV + point, n + 1)
+        }
+        map.toIterator
+      }.reduceByKeyLocally {
+        // sum the accumulation and the count in the all partition
+        case ((p1, n1), (p2, n2)) => (p1 + p2, n1 + n2)
+      }.map { case ((idx: Int, (center: BV[Double], counts: Int))) =>
+        center :/ counts.toDouble
+      }
+
+      val normSum = centers.map(v => breezeNorm(v, 2.0)).sum
+      val newNormSum = newCenters.map(v => breezeNorm(v, 2.0)).sum
+      error = math.abs((normSum - newNormSum) / normSum)
+      centers = newCenters.toArray
+      numIter += 1
+
+      logInfo(s"${numIter} iterations is finished" +
+          s" for ${System.currentTimeMillis() - startTimeOfIter}" +
+          s" at ${getClass}.split")
+    }
+
+    val vectors = centers.map(center => Vectors.fromBreeze(center))
+    val nodes = centers.size match {
+      case 1 => Array(new ClusterTree(vectors(0), data))
+      case 2 => {
+        val closest = data.map(p => (ClusterTree.findClosestCenter(metric)(centers)(p), p))
+        centers.zipWithIndex.map { case (center, i) =>
+          val subData = closest.filter(_._1 == i).map(_._2)
+          subData.cache
```

Comment:
Is it correct that caching every split (but then unpersisting) like this can potentially end up caching twice the size of the original data set? If so, that should be made very clear in the documentation, as it may affect the user's choice of what size data sets to run this one (and their hardware or instance provisioning).